File: examples/custom_ntrip.py
import socket
import base64

def connect_rtk2go_ntrip(server: str, port: int, mountpoint: str, username: str, password: str):
    """
    Connects to RTK2Go NTRIP caster and yields RTCM data chunks.

    Args:
        server (str): NTRIP caster (e.g., 'rtk2go.com')
        port (int): Caster port (usually 2101)
        mountpoint (str): RTK2Go mountpoint name (case-sensitive)
        username (str): Your RTK2Go email
        password (str): Password (typically 'none')

    Yields:
        bytes: Received RTCM data chunks
    """
    # Prepare HTTP Basic Auth header
    auth_str = f"{username}:{password}"
    auth_b64 = base64.standard_b64encode(auth_str.encode('utf-8')).decode('utf-8')

    # Construct NTRIP GET header
    request_header = (
        f"GET /{mountpoint} HTTP/1.0\r\n"
        "User-Agent: NTRIP PythonClient\r\n"
        "Accept: */*\r\n"
        f"Authorization: Basic {auth_b64}\r\n"
        "Connection: close\r\n\r\n"
    )

    # Create the TCP socket and connect
    sock = socket.create_connection((server, port))
    sock.sendall(request_header.encode('utf-8'))

    # Read and yield RTCM data
    try:
        while True:
            data = sock.recv(4096)  # Adjust buffer size as needed
            if not data:
                break
            yield data
    finally:
        sock.close()

for chunk in connect_rtk2go_ntrip('rtk2go.com', 2101, 'VMAX-LAND-1', 'hoppingturtles@proton.me', 'none'):
    # Do something with RTCM chunk, e.g., send to GNSS receiver
    print(chunk)

File: examples/custom_ntrip.rs
use base64::{Engine as _, engine::general_purpose};
use std::io::{Read, Write};
use std::net::TcpStream;

fn connect_rtk2go_ntrip(
    server: &str,
    port: u16,
    mountpoint: &str,
    username: &str,
    password: &str,
) -> std::io::Result<()> {
    // Connect to the RTK2GO NTRIP caster
    let mut stream = TcpStream::connect((server, port))?;

    // Prepare the auth header
    let auth = format!("{}:{}", username, password);
    let auth_b64 = general_purpose::STANDARD.encode(auth);

    // Construct NTRIP request header
    let request = format!(
        "GET /{} HTTP/1.0\r\n\
        User-Agent: NTRIP RustClient\r\n\
        Accept: */*\r\n\
        Authorization: Basic {}\r\n\
        Connection: close\r\n\r\n",
        mountpoint, auth_b64
    );

    stream.write_all(request.as_bytes())?;

    // Read and print incoming RTCM data chunks
    let mut buf = [0u8; 4096];
    loop {
        let n = stream.read(&mut buf)?;
        if n == 0 {
            break;
        }
        // Here you can process or forward the data. We'll just print length for this example:
        println!("Received {} bytes of RTCM data", n);
    }

    Ok(())
}

fn main() {
    // Replace with your RTK2go settings
    let server = "rtk2go.com";
    let port = 2101;
    let mountpoint = "VMAX-LAND-1";
    let username = "hoppingturtles@proton.me";
    let password = "none"; // typically 'none' for RTK2GO

    if let Err(e) = connect_rtk2go_ntrip(server, port, mountpoint, username, password) {
        eprintln!("NTRIP connection failed: {}", e);
    }
}
